### PR TITLE
feat: Record requests from the same test into one file

### PR DESCRIPTION
**Pull request by hkt74**
Remove the previous recorded request shaSum, now requests from the same test will be recored in the same file in sequence

### CODE REVIEW

File: internal/record/recording_https_proxy.go [hunk 1]
```diff
@@ -178,17 +161,40 @@ func (r *RecordingHTTPSProxy) proxyRequest(w http.ResponseWriter, req *http.Requ
...
 	recordedResponse, err := store.NewRecordedResponse(resp, body)
 	if err != nil {
 		return err
 	}
+	recordPath := filepath.Join(r.recordingDir, fileName+".http.log")
+	shaSum := recReq.ComputeSum()
 
-	recordedResponse.Body = r.redactor.Bytes(recordedResponse.Body)
+	// Open the file in append mode, create it if it doesn't exist.
+	file, err := os.OpenFile(recordPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
```
Comment (amirh) [inline internal/record/recording_https_proxy.go:173]: If the Test-Name header isn't set this still is still content hash right? in that case we may have cross-test collisions? previously it wasn't a problem because same hash meant same request (, same previous requests, ) and same response, but now that we're appending it might be a problem?

File: internal/record/recording_https_proxy.go [hunk 2]
```diff
@@ -178,17 +161,40 @@ func (r *RecordingHTTPSProxy) proxyRequest(w http.ResponseWriter, req *http.Requ
...
+	if err != nil {
+		return err
+	}
 
-	recordPath := filepath.Join(r.recordingDir, fileName+".resp")
 	fmt.Printf("Writing response to: %s\n", recordPath)
-	err = os.WriteFile(recordPath, []byte(recordedResponse.Serialize()), 0644)
+    recordedResponse.Body = r.redactor.Bytes(recordedResponse.Body)
+	serializedResp := recordedResponse.Serialize()
+	_, err = file.WriteString(fmt.Sprintf("\n%s.resp %d\n", shaSum, len(serializedResp)))
```
Comment (amirh) [inline internal/record/recording_https_proxy.go:193]: Why do we need to write `<shaSum>.resp` here?

File: internal/replay/replay_http_server.go [hunk 3]
```diff
@@ -122,14 +123,57 @@ func (r *ReplayHTTPServer) createRedactedRequest(req *http.Request) (*store.Reco
...
+		parts := strings.Fields(trimmedLine)
+		if len(parts) != 2 {
+			continue
+		}
+
+		fileKey := parts[0]
+		sizeStr := parts[1]
+
+		// 3. Check if the key on this line matches our target shaSum.
+		if fileKey == expectedKey {
```
Comment (amirh) [inline internal/replay/replay_http_server.go:158]: What happens when we're trying to reply the second response from a file?  do we need to skip size bytes anyway here?

File: internal/record/recording_https_proxy.go (hunk 2 above)
Comment (hkt74) [reply internal/record/recording_https_proxy.go:199]: I use `<shaSum>.resp` as delimiter to parse the replay response from the replay file

see an example replay log https://gist.github.com/hkt74/1ffa7b17bcf45b33d8b1a9d806a69784

File: internal/store/store.go [hunk 4]
```diff
@@ -102,7 +101,8 @@ func (r *RecordedRequest) GetRecordingFileName() (string, error) {
 		return "", fmt.Errorf("test name: %s contains illegal sequence '../'", testName)
 	}
 	if testName != "" {
-		return testName, nil
+		fileName := strings.ReplaceAll(testName, " ", "_")
```
Comment (amirh) [inline internal/store/store.go:104]: Could this potentially lead to name conflict? e.g a test named "print_" and a test named "print " writing to the same file?

File: internal/record/recording_https_proxy.go [hunk 5]
```diff
@@ -178,21 +168,49 @@ func (r *RecordingHTTPSProxy) proxyRequest(w http.ResponseWriter, req *http.Requ
...
 	}
+	recordPath := filepath.Join(r.recordingDir, fileName+".http.log")
 
-	recordedResponse.Body = r.redactor.Bytes(recordedResponse.Body)
+	var fileMode int
+	if fileName == shaSum {
+		// This is a single-request file, so overwrite it completely.
+		fileMode = os.O_TRUNC
+	} else {
+		// This file groups the requests by test-name, so append to it.
```
Comment (amirh) [inline internal/record/recording_https_proxy.go:183]: What's the story for re-running a test (e.g I changed the code and I want to re-record).
With current code it'll end up appending to the existing file and the errors may be confusing?

File: internal/replay/replay_http_server.go (hunk 3 above)
Comment (amirh) [reply internal/replay/replay_http_server.go:162]: Followed up f2f, I still think swe should skip the entire message if it's not a match.

**Review by amirh — CHANGES_REQUESTED** [review]

File: internal/replay/replay_http_server.go (hunk 3 above)
Comment (hkt74) [reply internal/replay/replay_http_server.go]: updated to skip not matched message entiredly

File: internal/record/recording_https_proxy.go (hunk 1 above)
Comment (hkt74) [reply internal/record/recording_https_proxy.go]: introduced seenMap to reset the recording when first time seen a request for a given test

File: internal/store/store.go (hunk 4 above)
Comment (hkt74) [reply internal/store/store.go:104]: sorry I misunderstood your original comment (or I might reply to a wrong thread).

Jasmine allows duplicate test name, so there could be conflict replay file for multiple tests.

1) if users run record together for 2 tests, both requests and response will be recorded in the same file. Which is strange but still works. Test sever itself doesn't know they are from 2 tests.
2) if users run record for 1 test and overwrite the same file for another test, the other test will fail, I think in this case users can notice there are duplicate test names, maybe not obvious and can be figured out quickly.

I could update the code to skip replace for head or tail characters, but seems over complicate the logic and cannot resolve the duplicate tests, I am leaning towards keep it as is and expect users to set the test name properly.

**Review by amirh — APPROVED** [review]
LGTM

File: internal/record/recording_https_proxy.go [hunk 6]
```diff
@@ -95,17 +101,18 @@ func (r *RecordingHTTPSProxy) handleRequest(w http.ResponseWriter, req *http.Req
...
-	err = r.recordResponse(resp, fileName, respBody)
-
+	shaSum := recReq.ComputeSum()
+	err = r.recordResponse(recReq, resp, fileName, shaSum, respBody)
 	if err != nil {
 		fmt.Printf("Error recording response: %v\n", err)
 		http.Error(w, fmt.Sprintf("Error recording response: %v", err), http.StatusInternalServerError)
 		return
 	}
+	r.prevRequestSHA = shaSum
```
[the remark this answers is not in the document]
Comment (hkt74) [reply internal/record/recording_https_proxy.go]: Done
